homeassistant/components/octoprint.py

```diff
@@ -4,72 +4,191 @@
+    """Wrap value in list if it is not one."""
+    if value is None or (isinstance(value, dict) and not value):
+        return []
+    return value if isinstance(value, list) else [value]
+
+
 CONFIG_SCHEMA = vol.Schema({
-    DOMAIN: vol.Schema({
-        vol.Required(CONF_API_KEY): cv.string,
+    DOMAIN: vol.All(ensure_list, [vol.Schema({
+        vol.Required(CONF_NAME): cv.string,
         vol.Required(CONF_HOST): cv.string,
+        vol.Optional(CONF_PORT): cv.string,
+        vol.Optional(CONF_SSL, default=False): cv.boolean,
+        vol.Required(CONF_API_KEY): cv.string,
         vol.Optional(CONF_NUMBER_OF_TOOLS, default=0): cv.positive_int,
         vol.Optional(CONF_BED, default=False): cv.boolean
-    }),
+    })])
 }, extra=vol.ALLOW_EXTRA)
 
 
-def setup(hass, config):
+@asyncio.coroutine
+def async_setup(hass, config):
     """Set up the OctoPrint component."""
-    base_url = 'http://{}/api/'.format(config[DOMAIN][CONF_HOST])
-    api_key = config[DOMAIN][CONF_API_KEY]
-    number_of_tools = config[DOMAIN][CONF_NUMBER_OF_TOOLS]
-    bed = config[DOMAIN][CONF_BED]
-
-    hass.data[DOMAIN] = {"api": None}
-
-    try:
-        octoprint_api = OctoPrintAPI(base_url, api_key, bed, number_of_tools)
-        hass.data[DOMAIN]["api"] = octoprint_api
-        octoprint_api.get('printer')
-        octoprint_api.get('job')
-    except requests.exceptions.RequestException as conn_err:
-        _LOGGER.error("Error setting up OctoPrint API: %r", conn_err)
+    if DOMAIN not in hass.data:
+        octoprints = hass.data[DOMAIN] = {}
+
+    @asyncio.coroutine
+    def async_setup_octoprint(octo_config):
+        """Set up an Octoprint service."""
+        from octoclient import OctoClient
+
+        name = octo_config[CONF_NAME]
+        _LOGGER.debug("Configuring Octoprint %s", octo_config[CONF_NAME])
+
+        schema = 'http'
+        if CONF_SSL in octo_config:
+            if octo_config[CONF_SSL] is True:
+                schema = 'https'
+
+        if CONF_PORT in octo_config:
+            if octo_config[CONF_HOST].find('/'):
+                host = octo_config[CONF_HOST].split('/', 1)[0]
+                path = octo_config[CONF_HOST].split('/', 1)[1]
+            else:
+                host = octo_config[CONF_HOST]
+                path = ''
+            url = '{}:{}/{}'.format(
+                host,
+                octo_config[CONF_PORT],
+                path
+            )
+        else:
+            url = '{}'.format(octo_config[CONF_HOST])
+
+        base_url = '{}://{}'.format(schema, url)
 
+        try:
+            octoprint = OctoClient(
+                url=base_url,
+                apikey=octo_config[CONF_API_KEY]
+            )
+            octoprint_api = OctoprintHandle(
+                name,
+                octoprint,
+                octo_config[CONF_NUMBER_OF_TOOLS],
+                octo_config[CONF_BED]
+            )
+            octoprint_api.get("printer")
```

Comment:
This is still calling a method that is doing I/O, within an async context, ie we're in a coroutine here.

The library isn't using asyncio. I suggest moving to a library that uses aiohttp.